AX: Move live region announcements into webkit behind feature flag

hoffmanjoshua opened the pull request:
#### c148f97f417f7724d88b029ab049b06c27468dd7
<pre>
AX: Move live region announcements into webkit behind feature flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=302716">https://bugs.webkit.org/show_bug.cgi?id=302716</a>
<a href="https://rdar.apple.com/164972478">rdar://164972478</a>

Reviewed by Tyler Wilcock.

This patch consists of the initial implementation of bringing live region announcements into
WebKit (guarded behind a runtime flag). In doing so, we can streamline how live regions
function across different ATs.

There are a few pieces to this patch:
- AXLiveRegionManager: this encapsulates the logic to compute announcements, changes, and
keep track of live regions. Each AXObjectCache has one AXLiveRegionManager, which is also
an indicator that the feature flag is enabled. Live regions are stored in the manager as
an array of &quot;snapshots&quot;, which are used to calculate changes when a live region changes.
- AXObjectCache changes: This is where we initialize the manager, and route node additions/
removals so that they are registered properly by the manager. The cache is also responsible
for posting the announcement notifications, using information computed by the manager.
- IPC (relayLiveRegionNotification): On iOS, we need to post notifications from the UI
process, rather than the web content process. These new IPC messages forward the information
needed to build an announcement notification in the UI process.

When this runtime flag is enabled, AXLiveRegionChanged and AXLiveRegionCreated notifications
will no longer be sent.

Several new tests are added to verify this initial behavior.

Tests: accessibility/ios-simulator/live-regions/live-region-types.html
       accessibility/mac/live-regions/live-region-alert-added.html
       accessibility/mac/live-regions/live-region-alert-on-initial-load.html
       accessibility/mac/live-regions/live-region-dynamically-added.html
       accessibility/mac/live-regions/live-region-hidden-content.html
       accessibility/mac/live-regions/live-region-removals.html
       accessibility/mac/live-regions/live-region-types.html
       accessibility/mac/live-regions/live-region-with-atomic.html

* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-types-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-types.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp: Added.
(WebCore::AXLiveRegionManager::AXLiveRegionManager):
(WebCore::debugDescriptionForSnapshot):
(WebCore::AXLiveRegionManager::registerLiveRegion):
(WebCore::stringToLiveRegionStatus):
(WebCore::stringToLiveRegionRelevant):
(WebCore::AXLiveRegionManager::handleLiveRegionChange):
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):
(WebCore::AXLiveRegionManager::textForObject const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
(WebCore::AXLiveRegionManager::postAnnouncementForChange):
* Source/WebCore/accessibility/AXLiveRegionManager.h: Added.
(WebCore::AXLiveRegionManager::unregisterLiveRegion):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::initializeLiveRegionManager):
(WebCore::AXObjectCache::postLiveRegionNotification):
(WebCore::AXObjectCache::postLiveRegionChangeNotification):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
(WebCore::AXObjectCache::deferSortForNewLiveRegion):
(WebCore::AXObjectCache::sortedLiveRegions):
(WebCore::AXObjectCache::initializeSortedIDLists):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::relayLiveRegionNotification const):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXRemovedText):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayLiveRegionNotification):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::relayLiveRegionNotification):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebChromeClientIOS.mm:
(WebKit::WebChromeClient::relayLiveRegionNotification const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::relayAriaNotifyNotification):
(WebKit::WebPage::relayLiveRegionNotification):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:

Canonical link: <a href="https://commits.webkit.org/303361@main">https://commits.webkit.org/303361@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65835b1dcff695135fb1456e9a570b0b608dcce9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139603 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84001 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a1c1e68c-e2b7-4238-8a20-2556086ddb10) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4342 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100972 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/01087531-0529-4b89-b9b0-b836634cb993) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81763 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c36cc779-f445-4b36-9f53-1ccf2297f812) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/131437 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82822 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124151 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111893 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142249 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130595 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4250 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37007 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109344 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4331 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109518 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27754 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3228 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114583 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57494 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4304 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32977 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163562 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4135 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67750 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42511 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4395 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4263 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load.html:
```diff
@@ -0,0 +1,43 @@
+<div id="alert-region" role="alert">Important announcement</div>
+
+<script>
+var output = "This tests that when an alert (implicit live region) is initialized, it is announced.\n\n";    
```
Do we want a test for an alert being added to the page after initial load?

hoffmanjoshua replied:
Added!

minorninth commented on LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added.html:
```diff
@@ -0,0 +1,53 @@
+    document.getElementById('body').appendChild(newRegion);
+ 
+    setTimeout(async function() {
+        await sleep(500);
```
Rather than sleeping for 500 ms, could you do a waitFor() on accessibleElementById on the live region? Once the live region appears in the accessibility tree you should be able to modify it and know it will be announced.

hoffmanjoshua replied:
Good idea! Fixed this in all the tests

minorninth commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,317 @@
+    return result;
+}
+
+void AXLiveRegionManager::handleLiveRegionChange(AccessibilityObject& object, bool speakAll)
```
How about an enum instead of bool here? Values like AnnounceEntireRegion vs AnnounceChangesOnly

Also let's prefer verbs like "announce" because live regions are rendered in braille too, not just speech

hoffmanjoshua replied:
Updated this to an enum

minorninth commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,317 @@
+{
+    auto axID = object.objectID();
+
+    // This is a new live region, don't speak it upon registering.
```
Should this say "if" this is a new live region?

minorninth commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,317 @@
+                removalString.append(object.text);
+            }
+        }
+        strings.append(removalString.toString());
```
It looks like it's possible for diff.removed to be nonempty, but all of the text is empty, so you may not want to append to strings in that case

hoffmanjoshua replied:
WTFSTring::append is a no-op if the string param is empty, so we should be okay to leave this as-is

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-removals.html:
```diff
@@ -0,0 +1,60 @@
+<body>
+
+<ul id="team-list" aria-live="polite" aria-relevant="removals" style="list-style-type: none;">
+    <li>Giants</li>
+    <li>Cubs</li>
+    <li>Orioles</li>
+    <li>Tigers</li>
```
No Brewers???

hoffmanjoshua replied:
Go cubbies!

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-removals.html:
```diff
@@ -0,0 +1,60 @@
+        if (notificationCount == 1)
+            testFinished = true;
+    }
+	
```
Extra newline

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-hidden-content.html:
```diff
@@ -0,0 +1,55 @@
+
+        testFinished = true;
+    }
+	
```
Extra newline

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-with-atomic.html:
```diff
@@ -0,0 +1,66 @@
+            document.getElementById("state-list").appendChild(newState);
+        } else if (notificationCount == 2)
+            testFinished = true;
+
```
Extra newline

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-with-atomic.html:
```diff
@@ -0,0 +1,66 @@
+            testFinished = true;
+
+    }
+    
```
Extra newline

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+{
+}
+
+static UNUSED_FUNCTION String debugDescriptionForSnapshot(LiveRegionSnapshot snapshot)
```
Thanks for adding this 🙏

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+    if (snapshot.liveRegionRelevant.isEmpty())
+        result.append("(default: additions text)"_s);
+    else {
+        bool first = true;
```
I think WebKit prefers bools to start with `is` / `has` / etc

hoffmanjoshua replied:
Fixed!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+
+void AXLiveRegionManager::registerLiveRegion(AccessibilityObject& object, bool speakIfNecessary)
+{
+    auto axID = object.objectID();
```
Any need for this local? Seems like we only use it once

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+
+void AXLiveRegionManager::handleLiveRegionChange(AccessibilityObject& object, AnnouncementContents contents)
+{
+    auto axID = object.objectID();
```
Any need for this local? Seems like we only use it once.

hoffmanjoshua replied:
Ah yeah we don't need this anymore, removed this and the other occurrence

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+    auto axID = object.objectID();
+
+    // If this is a new live region, don't speak it upon registering.
+    auto it = m_liveRegions.find(axID);
```
I think WebKit's coding style prefers spelling out variable names. Maybe `iterator`?

hoffmanjoshua replied:
Fixed!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+        if (RefPtr childObject = dynamicDowncast<AccessibilityObject>(child.get()))
+            buildObjectList(*childObject, objects);
+    }
+}
```
Since this is only used within a single function `buildLiveRegionSnapshot`, can we just make it a lambda that lives within that function? Or is there some reason you feel this should be separated out this way? It could be nice for that function to own all of the logic so you don't need to track down this other method

hoffmanjoshua replied:
Moved to a lambda, great call!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,318 @@
+    // Treat atomic objects as one object, so when they change the entire subtree is announced.
+    if (object.liveRegionAtomic()) {
+        auto text = textForObject(object);
+        objects.append({ object.objectID(), text });
```
We can WTFMove(text) here, or better yet not bother with a local variable?

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,312 @@
+void AXLiveRegionManager::unregisterLiveRegion(AXID axID)
+{
+    m_liveRegions.remove(axID);
+}
```
Can this be inlined in the header?

hoffmanjoshua replied:
Moved!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,312 @@
+    Vector<String> strings = string.split(" "_s);
+    OptionSet<LiveRegionRelevant> result;
+    for (const auto& attribute : strings) {
+        if (attribute == "additions")
```
The spec technically allows authors to pass any case string (not just lowercase).

https://w3c.github.io/html-aria#case-sensitivity

Wonder if we should add a FIXME with this spec link?

hoffmanjoshua replied:
Fixed this now so it wouldn't be an issue by converting the whole string to lowercase.

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,312 @@
+
+    // If an object has unignored children, there isn't a need to include it in the snapshot since the children will return YES.
+    auto children = object.unignoredChildren();
+    if (!children.isEmpty())
```
It is much more efficient to use the `AXCoreObject::firstUnignoredChild` here

hoffmanjoshua replied:
Fixed ✅

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.h:
```diff
@@ -0,0 +1,90 @@
+struct LiveRegionSnapshot {
+    Vector<LiveRegionObject> objects;
+    LiveRegionStatus liveRegionStatus;
+    OptionSet<LiveRegionRelevant> liveRegionRelevant;
```
Should we have default initializer values for these, e.g. `{ ...default... };`

hoffmanjoshua replied:
Updated!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,312 @@
+
+    LiveRegionDiff diff;
+
+    // Build a map of old objects for lookup. As we match them with new objects, we'll remove them.
+    // Whatever remains unmatched at the end represents removals.
+    HashMap<AXID, String> unmatchedOldObjects;
+    for (auto& object : oldObjects)
+        unmatchedOldObjects.set(object.objectID, object.text);
```
It feels strange to basically copy this vector into a hash map. Is it possible for `computeChanges` to take a HashMap reference? Or is there a reason we're using a Vector of structs that are key-value pairs? Maybe I just don't understand all the complexities of the algorithm

hoffmanjoshua replied:
Yeah, I explored a few alternatives to this, but because we need to maintain ordering in these snapshots, this was the approach I settled on. If we made this take in a HashMap, and also stored a hash map in the snapshot, we would lose ordering or have to increase our memory footprint by adding a sorted array of AXIDs to keep track.

Some other algorithms I explored have a higher complexity (for example, keeping a list of "seen" objects) due to more expensive lookups.

Thankfully here while we are creating a new HashMap, the strings aren't being copied (they are just references) so we shouldn't be increasing the memory footprint by much.

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,312 @@
+        return;
+
+    String announcementText = computeAnnouncement(newSnapshot, diff);
+
```
This newline feels unnecessary

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -593,22 +598,27 @@ String AXTextMarkerRange::toString(IncludeListMarkerText includeListMarkerText)
             return { };
 
-        TextIterator it = TextIterator(*range, { TextIteratorBehavior::IgnoresFullSizeKana });
+        OptionSet<TextIteratorBehavior> textIteratorBehaviors = { TextIteratorBehavior::IgnoresFullSizeKana };
```
I think since we're spelling out the type, shortening the name to something like `behaviors` is fine

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,331 @@
+    if (string == "assertive")
+        return LiveRegionStatus::Assertive;
+    if (string == "polite")
+        return LiveRegionStatus::Polite;
```
These probably also need to be converted to lower case

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,331 @@
+        for (auto& child : object.unignoredChildren()) {
+            if (RefPtr childObject = dynamicDowncast<AccessibilityObject>(child.get()))
+                buildObjectList(*childObject);
+        }
```
IMO we should just downcast here. Crashing is probably the best move if an AccessibilityObject ever has an AXIsolatedObject child, and I think it would be more performant.

```cpp
for (auto& child : object.unignoredChildren())
    buildObjectList(downcast<AccessibilityObject>(child.get()), objects);
```

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,331 @@
+
+    // Build a map of old objects for lookup. As we match them with new objects, we'll remove them.
+    // Whatever remains unmatched at the end represents removals.
+    HashMap<AXID, String> unmatchedOldObjects;
```
Can we improve performance by using `reserveInitialCapacity` since we know how many keys we'll be adding?

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -0,0 +1,331 @@
+            if (!object.text.isEmpty()) {
+                removalString.append(' ');
+                removalString.append(object.text);
+                characterCount += object.text.length() + 1;
```
Can we add a comment explaining what the `+ 1` is?

twilco reviewed (approved):
r+ with comments from this latest round addressed

hoffmanjoshua committed:
AX: Move live region announcements into webkit behind feature flag
https://bugs.webkit.org/show_bug.cgi?id=302716
rdar://164972478

Reviewed by Tyler Wilcock.

This patch consists of the initial implementation of bringing live region announcements into
WebKit (guarded behind a runtime flag). In doing so, we can streamline how live regions
function across different ATs.

There are a few pieces to this patch:
- AXLiveRegionManager: this encapsulates the logic to compute announcements, changes, and
keep track of live regions. Each AXObjectCache has one AXLiveRegionManager, which is also
an indicator that the feature flag is enabled. Live regions are stored in the manager as
an array of "snapshots", which are used to calculate changes when a live region changes.
- AXObjectCache changes: This is where we initialize the manager, and route node additions/
removals so that they are registered properly by the manager. The cache is also responsible
for posting the announcement notifications, using information computed by the manager.
- IPC (relayLiveRegionNotification): On iOS, we need to post notifications from the UI
process, rather than the web content process. These new IPC messages forward the information
needed to build an announcement notification in the UI process.

When this runtime flag is enabled, AXLiveRegionChanged and AXLiveRegionCreated notifications
will no longer be sent.

Several new tests are added to verify this initial behavior.

Tests: accessibility/ios-simulator/live-regions/live-region-types.html
       accessibility/mac/live-regions/live-region-alert-added.html
       accessibility/mac/live-regions/live-region-alert-on-initial-load.html
       accessibility/mac/live-regions/live-region-dynamically-added.html
       accessibility/mac/live-regions/live-region-hidden-content.html
       accessibility/mac/live-regions/live-region-removals.html
       accessibility/mac/live-regions/live-region-types.html
       accessibility/mac/live-regions/live-region-with-atomic.html

* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-types-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-types.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp: Added.
(WebCore::AXLiveRegionManager::AXLiveRegionManager):
(WebCore::debugDescriptionForSnapshot):
(WebCore::AXLiveRegionManager::registerLiveRegion):
(WebCore::stringToLiveRegionStatus):
(WebCore::stringToLiveRegionRelevant):
(WebCore::AXLiveRegionManager::handleLiveRegionChange):
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):
(WebCore::AXLiveRegionManager::textForObject const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
(WebCore::AXLiveRegionManager::postAnnouncementForChange):
* Source/WebCore/accessibility/AXLiveRegionManager.h: Added.
(WebCore::AXLiveRegionManager::unregisterLiveRegion):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::initializeLiveRegionManager):
(WebCore::AXObjectCache::postLiveRegionNotification):
(WebCore::AXObjectCache::postLiveRegionChangeNotification):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
(WebCore::AXObjectCache::deferSortForNewLiveRegion):
(WebCore::AXObjectCache::sortedLiveRegions):
(WebCore::AXObjectCache::initializeSortedIDLists):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::relayLiveRegionNotification const):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXRemovedText):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayLiveRegionNotification):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::relayLiveRegionNotification):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebChromeClientIOS.mm:
(WebKit::WebChromeClient::relayLiveRegionNotification const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::relayAriaNotifyNotification):
(WebKit::WebPage::relayLiveRegionNotification):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:

Canonical link: https://commits.webkit.org/303361@main
